File: shell/gnome/src/popup/usageRows.js
import Clutter from 'gi://Clutter';
import St from 'gi://St';
import { dayTitle } from '../dates.js';
import { _ } from '../i18n.js';
import { labelText } from '../labels.js';
import { compactTokensText, exactSpendLine, exactTokens, money, spendLine, usd } from '../numbers.js';
import { seriesKey } from '../providers.js';
import { sameValues } from '../sameValues.js';
import { column, label, row } from '../widgets.js';
import { MASK } from './mask.js';
import { modelTooltip } from './modelTooltip.js';
import { showsSpend } from './sectionShape.js';

const TREND_DAYS = 30;
const TREND_HEIGHT = 18;
const TREND_STUB = 2;
const TREND_MIN_SHARE = 0.18;

function lastDays(daily) {
    const days = daily.slice(-TREND_DAYS);
    const padding = Array.from({ length: TREND_DAYS - days.length }, () => ({
        date: '',
        totalTokens: 0,
        costMicros: 0,
        partial: false,
    }));
    return [...padding, ...days];
}

function barHeight(value, peak) {
    if (value <= 0 || peak <= 0) return TREND_STUB;
    return Math.max(Math.round(TREND_HEIGHT * TREND_MIN_SHARE), Math.round((TREND_HEIGHT * value) / peak));
}

function dayTooltip(day) {
    if (day.date === '') return null;
    const actor = column({ style_class: 'headroom-tip' });
    actor.add_child(label(dayTitle(day.date), 'headroom-tip-title', { x_align: Clutter.ActorAlign.START }));
    const partial = day.partial ? _(' · some models unpriced') : '';
    const figures =
        day.totalTokens === 0 ? _('No usage') : `${compactTokensText(day.totalTokens)} · ${usd(day.costMicros)}`;
    actor.add_child(label(`${figures}${partial}`, 'headroom-tip-figure', { x_align: Clutter.ActorAlign.START }));
    return actor;
}

export class TrendRow {
    constructor(ctx, usage) {
        this._days = lastDays(usage.daily);
        this._tokens = [];
        this.actor = row({ style_class: 'headroom-text-row' });
        this.actor.add_child(label(_('Usage Trend'), 'headroom-value-label', { x_expand: true }));
        const strip = row({ style_class: 'headroom-trend', y_align: Clutter.ActorAlign.CENTER });
        this._bars = this._days.map((_day, index) => {
            const slot = new St.Widget({ style_class: 'headroom-trend-slot', layout_manager: new Clutter.BinLayout() });
            const bar = new St.Widget({
                style_class: 'headroom-trend-bar',
                y_expand: true,
                y_align: Clutter.ActorAlign.END,
            });
            slot.add_child(bar);
            ctx.tooltips.attach(slot, () => (ctx.masked ? null : dayTooltip(this._days[index])));
            strip.add_child(slot);
            return bar;
        });
        this.actor.add_child(strip);
        this.update(usage);
    }

    update(usage) {
        this._days = lastDays(usage.daily);
        const tokens = this._days.map(day => day.totalTokens);
        if (sameValues(tokens, this._tokens)) return;
        this._tokens = tokens;
        const peak = Math.max(...tokens);
        tokens.forEach((value, index) => {
            this._bars[index].style = `height: ${barHeight(value, peak)}px;`;
        });
    }
}

function totalsTooltip(ctx, title, provider, totals) {
    if (totals.totalTokens === 0 || ctx.masked) return null;
    return modelTooltip(title, totals, { series: seriesKey(provider) }) ?? exactSpendLine(totals);
}

function balanceTitle(balance) {
    return labelText(balance.label);
}

function balanceValue(balance) {
    if (balance.kind === 'usd' && balance.usdMicros !== null) return usd(balance.usdMicros);
    if (balance.kind === 'money' && balance.currency !== null && balance.micros !== null)
        return money(balance.currency, balance.micros);
    if (balance.kind === 'count' && balance.value !== null)
        return `${exactTokens(balance.value)} ${balance.unit ?? ''}`.trim();
    return _('No data');
}

export class ExtraRows {
    constructor(ctx, account) {
        this._ctx = ctx;
        this._account = account;
        this.actor = column({ style_class: 'headroom-expanded', x_expand: true });
        this._values = [...this._spendEntries(), ...this._balanceEntries()].map(entry => this._valueRow(entry));
        this.update(account);
    }

    update(account) {
        this._account = account;
        const texts = [...this._spendEntries(), ...this._balanceEntries()].map(entry =>
            this._ctx.masked ? MASK : entry.value()
        );
        texts.forEach((text, index) => (this._values[index].text = text));
    }

    _spendEntries() {
        if (!showsSpend(this._ctx, this._account)) return [];
        const provider = this._account.providerName;
        const series = this._account.provider;
        return [
            ['today', _('Today')],
            ['yesterday', _('Yesterday')],
            ['last30Days', _('Last 30 Days')],
        ].map(([key, title]) => ({
            title,
            value: () => spendLine(this._account.usage[key]),
            tooltip: () => totalsTooltip(this._ctx, `${provider} · ${title}`, series, this._account.usage[key]),
        }));
    }

    _balanceEntries() {
        return this._account.balances.map((_balance, index) => ({
            title: balanceTitle(this._account.balances[index]),
            value: () => balanceValue(this._account.balances[index]),
            tooltip: null,
        }));
    }

    _valueRow(entry) {
        const actor = row({ style_class: 'headroom-text-row' });
        actor.add_child(label(entry.title, 'headroom-value-label', { x_expand: true }));
        const valueLabel = label('', 'headroom-value');
        if (entry.tooltip) {
            valueLabel.add_style_class_name('headroom-hover-chip');
            this._ctx.tooltips.attach(valueLabel, entry.tooltip);
        }
        actor.add_child(valueLabel);
        this.actor.add_child(actor);
        return valueLabel;
    }
}

File: crates/headroom-tray/src/ui/usage_rows.rs
use gtk::prelude::*;

use crate::dates::day_title;
use crate::i18n::Lang;
use crate::labels::{balance_title, balance_value};
use crate::numbers::{compact_tokens_text, exact_spend_line, spend_line, usd};
use crate::payload::{Account, Daily, Usage};
use crate::spend::{Period, TREND_HEIGHT, bar_height, breakdown_text, trend_days};
use crate::ui::context::Ctx;
use crate::ui::draw::{fill, rounded_top, set_color};
use crate::ui::widgets::{column, label, row};

const BAR_WIDTH: f64 = 4.0;
const BAR_GAP: f64 = 1.0;
const BAR_SLOTS: i32 = 30;
const COMPACT_TREND_HEIGHT: i32 = 14;

fn day_tooltip(lang: Lang, day: &Daily) -> String {
    let figures = if day.total_tokens == 0 {
        lang.tr("No usage").to_owned()
    } else {
        format!(
            "{} · {}",
            compact_tokens_text(lang, day.total_tokens),
            usd(day.cost_usd_micros)
        )
    };
    let partial = if day.partial {
        lang.tr(" · some models unpriced")
    } else {
        ""
    };
    format!("{}\n{figures}{partial}", day_title(lang, day.date))
}

fn strip_width() -> i32 {
    let width = f64::from(BAR_SLOTS) * BAR_WIDTH + f64::from(BAR_SLOTS - 1) * BAR_GAP;
    #[allow(
        clippy::cast_possible_truncation,
        reason = "a small constant pixel width"
    )]
    let pixels = width as i32;
    pixels
}

fn slot_tooltip(heights: &[(u64, Option<Daily>)], x: i32, lang: Lang) -> Option<String> {
    #[allow(
        clippy::cast_possible_truncation,
        clippy::cast_sign_loss,
        reason = "a small positive constant slot pitch"
    )]
    let pitch = (BAR_WIDTH + BAR_GAP) as usize;
    let slot = usize::try_from(x).ok()? / pitch;
    heights
        .get(slot)?
        .1
        .as_ref()
        .map(|day| day_tooltip(lang, day))
}

fn day_heights(usage: &Usage) -> Vec<(u64, Option<Daily>)> {
    let days = trend_days(&usage.daily);
    let peak = days
        .iter()
        .flatten()
        .map(|day| day.total_tokens)
        .max()
        .unwrap_or(0);
    days.iter()
        .map(|day| {
            (
                bar_height(day.map_or(0, |d| d.total_tokens), peak),
                day.cloned(),
            )
        })
        .collect()
}

fn trend_strip(ctx: &Ctx, usage: &Usage) -> gtk::DrawingArea {
    let heights = day_heights(usage);
    let area = gtk::DrawingArea::new();
    area.set_content_width(strip_width());
    let full = i32::try_from(TREND_HEIGHT).unwrap_or(18);
    area.set_content_height(if ctx.compact() {
        COMPACT_TREND_HEIGHT
    } else {
        full
    });
    area.set_valign(gtk::Align::Center);
    let color = ctx.tone_color(crate::payload::Tone::Good);
    let bars: Vec<u64> = heights.iter().map(|(height, _)| *height).collect();
    area.set_draw_func(move |_, cr, _, height| {
        set_color(cr, color);
        let scale = f64::from(height) / f64::from(full);
        for (index, bar) in bars.iter().enumerate() {
            #[allow(clippy::cast_precision_loss, reason = "30 slots and 18 px bars")]
            let (x, bar) = (index as f64 * (BAR_WIDTH + BAR_GAP), *bar as f64 * scale);
            rounded_top(cr, x, f64::from(height) - bar, BAR_WIDTH, bar, 1.0);
            fill(cr);
        }
    });
    area.set_has_tooltip(true);
    let lang = ctx.locale.lang;
    area.connect_query_tooltip(move |_, x, _, _, tooltip| {
        let text = slot_tooltip(&heights, x, lang);
        tooltip.set_text(text.as_deref());
        text.is_some()
    });
    area
}

pub fn trend_row(ctx: &Ctx, usage: &Usage) -> gtk::Box {
    let line = row(10, &["headroom-text-row"]);
    let title = label(ctx.locale.lang.tr("Usage Trend"), &["headroom-value-label"]);
    title.set_hexpand(true);
    line.append(&title);
    line.append(&trend_strip(ctx, usage));
    line
}

fn value_row(title: &str, value: &str, tooltip: Option<&str>) -> gtk::Box {
    let line = row(10, &["headroom-text-row"]);
    let name = label(title, &["headroom-value-label"]);
    name.set_hexpand(true);
    let figure = label(value, &["headroom-value"]);
    if let Some(text) = tooltip {
        figure.add_css_class("headroom-hover-chip");
        figure.set_tooltip_text(Some(text));
    }
    line.append(&name);
    line.append(&figure);
    line
}

fn spend_rows(ctx: &Ctx, account: &Account, usage: &Usage) -> Vec<gtk::Box> {
    let lang = ctx.locale.lang;
    Period::ALL
        .iter()
        .map(|period| {
            let totals = period.totals(usage);
            let title = period.row_title(lang);
            let heading = format!("{title} · {}", account.provider_name);
            let tooltip = (totals.tokens.total > 0).then(|| {
                let figures = (totals.cost_usd_micros, totals.tokens.total);
                breakdown_text(
                    lang,
                    &heading,
                    &totals.models,
                    totals.models_other.as_ref(),
                    figures,
                )
                .unwrap_or_else(|| exact_spend_line(lang, figures.0, figures.1, totals.partial))
            });
            let value = spend_line(lang, totals.cost_usd_micros, totals.tokens.total);
            value_row(title, &value, tooltip.as_deref())
        })
        .collect()
}

pub fn extra_rows(ctx: &Ctx, account: &Account, usage: Option<&Usage>) -> Option<gtk::Box> {
    let lang = ctx.locale.lang;
    let mut rows = Vec::new();
    if let Some(usage) = usage.filter(|_| ctx.display.show_account_spend) {
        rows.extend(spend_rows(ctx, account, usage));
    }
    rows.extend(account.balances.iter().filter_map(|balance| {
        let value = balance_value(lang, balance)?;
        Some(value_row(&balance_title(lang, balance), &value, None))
    }));
    if rows.is_empty() {
        return None;
    }
    let body = column(0, &[]);
    for line in rows {
        body.append(&line);
    }
    Some(body)
}
